server/app/views/admin/programs/DraftProgramBlockEditView.java

```diff
@@ -14,140 +13,248 @@
 
-  private final AdminLayout layout;
-  private final boolean featureFlagOptionalQuestions;
-
-  public static final String ENUMERATOR_ID_FORM_FIELD = "enumeratorId";
-  public static final String MOVE_QUESTION_POSITION_FIELD = "position";
   private static final String CREATE_BLOCK_FORM_ID = "block-create-form";
   private static final String CREATE_REPEATED_BLOCK_FORM_ID = "repeated-block-create-form";
   private static final String DELETE_BLOCK_FORM_ID = "block-delete-form";
 
+  // To avoid concurrency issues and duplication, collect all modals which this class
+  // and its children want to show in a synchronized Set.
+  private List<Modal> modals = Collections.synchronizedList(new ArrayList<Modal>());
+
+  private final boolean featureFlagOptionalQuestions;
+  private InputTag csrfTag;
+
   @Inject
-  public ProgramBlockEditView(AdminLayoutFactory layoutFactory, Config config) {
-    this.layout = checkNotNull(layoutFactory).getLayout(NavPage.PROGRAMS);
+  public DraftProgramBlockEditView(AdminLayoutFactory layoutFactory, Config config) {
+    super(layoutFactory, config);
     this.featureFlagOptionalQuestions = checkNotNull(config).hasPath("cf.optional_questions");
   }
 
+  @Override
   public Content render(
       Request request,
-      ProgramDefinition program,
+      ProgramDefinition programDefinition,
+      BlockForm blockForm,
       BlockDefinition blockDefinition,
       Optional<ToastMessage> message,
       ImmutableList<QuestionDefinition> questions) {
-    return render(
-        request,
-        program,
-        blockDefinition.id(),
-        new BlockForm(blockDefinition.name(), blockDefinition.description()),
-        blockDefinition,
-        blockDefinition.programQuestionDefinitions(),
-        message,
-        questions);
+
+    modals = Collections.synchronizedList(new ArrayList<Modal>());
+    csrfTag = makeCsrfTokenInputTag(request);
+    return super.render(request, programDefinition, blockForm, blockDefinition, message, questions);
   }
 
-  public Content render(
+  @Override
+  protected HtmlBundle createHtmlBundle(
       Request request,
       ProgramDefinition programDefinition,
-      long blockId,
       BlockForm blockForm,
       BlockDefinition blockDefinition,
-      ImmutableList<ProgramQuestionDefinition> blockQuestions,
-      Optional<ToastMessage> message,
       ImmutableList<QuestionDefinition> questions) {
-    InputTag csrfTag = makeCsrfTokenInputTag(request);
-    String title = String.format("Edit %s", blockDefinition.name());
 
+    return super.createHtmlBundle(request, programDefinition, blockForm, blockDefinition, questions)
+        .addMainContent(
+            questionBankPanel(
+                questions,
+                programDefinition,
+                blockDefinition,
+                QuestionBank.shouldShowQuestionBank(request)))
+        .addMainContent(addFormEndpoints(programDefinition.id(), blockDefinition.id()))
+        .addModals(modals);
```

Comment:
I believe you can compute modals here instead of having to add to it in `prepareContentForBlockPanel

I traced back up from where you populate modals and it appears only one is ever made and its with data available in this method.

I believe you can compute the modal with 

```
String blockUpdateAction =
        controllers.admin.routes.AdminProgramBlocksController.update(
                programDefinition.id(), blockDefinition.id())
            .url();
 Modal blockDescriptionEditModal = blockDescriptionModal(blockForm, blockUpdateAction);
    ```

Reply:
The problem isn't the creation, it's the fact that I need to get the button that triggers the modal from the modal.

See line 153

Is there a way of making a button trigger a modal without having access to it?

Reply:
Ah ok.  Given that a slightly better approach then I think is to create the modal in render() (instead of resetting the Modal set) and then use it on the appropriate button here.

We're then at least putting the creation at the top of the call chain and more knowingly creating it once and acknowledging that it's a single item and not a mutable list.

```
Optional<Modal> blockDescriptionEditModal = Optional.empty();

public render(....) {
  if(blockDescriptionEditModal.isEmpty()) {
    String blockUpdateAction = controllers.admin.routes.AdminProgramBlocksController.update( programDefinition.id(), blockDefinition.id()) .url();
   blockDescriptionEditModal = Optional.of(blockDescriptionModal(blockForm, blockUpdateAction));
}
```